### ISSUE TITLE
Single state label getting split into multiple states

### ISSUE THREAD

**Issue by joefowler** [usnistgov/mass#198]
**[Original report](https://bitbucket.org/joe_fowler/mass/issues/198) by Paul Szypryt (Bitbucket: [pszypryt](https://bitbucket.org/pszypryt), GitHub: [pszypryt](https://github.com/pszypryt)).**

----------------------------------------

It seems that in certain uses of the `states` argument in off file analysis, if you feed in a string rather than a list, the string can still get interpreted as multiple states. For example, in the following code I intend to plot the energy histogram for state ‘BG’, but instead I get histograms plotted for states ‘B' and 'G’, as shown in the attached image.  

```python
data.plotHist(np.arange(100,2500,1),'energy',states='BG',coAddStates=False)
```

The behavior seems to be different depending on whether `coAddStates` is set to `True` or `False`. There are likely other instances of this bug in other off file convenience functions.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

This sounds almost certainly like the rather ugly Python problem that arises because strings are iterable. So if you expect to loop over the elements of a list of strings or a tuple of strings, but the caller passes in a string instead, it still works! Sort of, because now each iteration is a 1-character string, the characters of the original. So the problem is that a string can behave as a collection of strings. Boo! Unfortunately, that’s the way Python semantics work.

If I have the right interpretation, then your problem is basically inevitable. The dependence on `coAddStates` is surely a result of looping over `states` only when the former is `True`, or at least in some different way.

Not sure what to suggest. Some ideas are:

1. Test the argument `states`. If it is a string, silently replace it with `(states, )` or `[states]. I suggest this one.`
2. Test the argument `states`. If it is a string, raise an error.
3. Take two arguments. `state` is expected to be a string, and `states` an iterable collection of them. Seems unpythonic and easy for a caller to mess up.
4. Rewrite Mass in a language that doesn’t treat a string as an iterable, or--just as good--doesn’t treat characters/runes/glyphs as length-1 strings.

‌

**Comment by joefowler** [comment]
**Original comment by Paul Szypryt (Bitbucket: [pszypryt](https://bitbucket.org/pszypryt), GitHub: [pszypryt](https://github.com/pszypryt)).**

----------------------------------------

I think Joe’s solution 1 would work best here, especially if the idea was that you could pass in a str if you just wanted to grab a single state.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Apparently [Guido himself tried to fix this problem](https://stackoverflow.com/questions/9168904/suppressing-treatment-of-string-as-iterable) back in 2006 and gave up, deciding it was too late.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Notice the function `mass.common.isstr(arg)` to test, in a Python 2/3-safe way, whether arg is a string-like object.

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

Bug is here: [https://bitbucket.org/joe\_fowler/mass/src/ac86a0cf446a3b3959b9a548d4332868e1981111/mass/off/channels.py#lines-237](https://bitbucket.org/joe_fowler/mass/src/ac86a0cf446a3b3959b9a548d4332868e1981111/mass/off/channels.py#lines-237)

This function handles it correctly: [https://bitbucket.org/joe\_fowler/mass/src/ac86a0cf446a3b3959b9a548d4332868e1981111/mass/off/channels.py#lines-435](https://bitbucket.org/joe_fowler/mass/src/ac86a0cf446a3b3959b9a548d4332868e1981111/mass/off/channels.py#lines-435)

Not sure if we should just fix it in one spot or try to make them share the code. Probably we can do the simplest thing.

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

fixed by PR #152

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Removing milestone: v0.7 (automated comment)
